main/src/library/Chain.flix

```diff
@@ -0,0 +1,865 @@
+            let sb = StringBuilder.new();
+            StringBuilder.appendString!(sb, "Chain#{");
+            Chain.foldLeft((acc, x) -> {
+                if (acc < 1)
+                    StringBuilder.appendString!(sb, "${x}")
+                else
+                    StringBuilder.appendString!(sb, ", ${x}");
+                acc + 1}, 0, xs);
+            StringBuilder.appendString!(sb, "}");
+            StringBuilder.toString(sb)
+        };
+        fn() as & Pure
+}
+
+instance Eq[Chain[a]] with Eq[a] {
+    pub def eq(xs: Chain[a], ys: Chain[a]): Bool = Chain.equals(xs, ys)
+}
+
+instance SemiGroup[Chain[a]] {
+    pub def combine(xs: Chain[a], ys: Chain[a]): Chain[a] = Chain.append(xs, ys)
+}
+
+instance Monoid[Chain[a]] {
+    pub def empty(): Chain[a] = Chain.empty()
+    pub def combine(xs: Chain[a], ys: Chain[a]): Chain[a] = Chain.append(xs, ys)
+}
+
+instance Functor[Chain] {
+    pub def map(f: a -> b & ef, xs: Chain[a]): Chain[b] & ef = Chain.map(f, xs)
+}
+
+instance Applicative[Chain] {
+    pub def point(x: a) : Chain[a] = Chain.singleton(x)
+    pub def ap(f: Chain[a -> b & ef], x: Chain[a]) : Chain[b] & ef = Chain.ap(f, x)
+}
+
+instance Monad[Chain] {
+    pub def flatMap(f: a -> Chain[b] & ef, x : Chain[a]) : Chain[b] & ef = Chain.flatMap(f, x)
+}
+
+instance Foldable[Chain] {
+    pub def foldLeft(f: (b, a) -> b & ef, s: b, xs: Chain[a]): b & ef = Chain.foldLeft(f, s, xs)
+    pub def foldRight(f: (a, b) -> b & ef, s: b, xs: Chain[a]): b & ef = Chain.foldRight(f, s, xs)
+}
+
+
+
+
+namespace Chain {
+
+    ///
+    /// A datatype for pattern matching on a chain (traversing left-to-right).
+    ///
+    pub enum ViewLeft[t] with Eq {
+        case NoneLeft
+        case SomeLeft(t, Chain[t])
+    }
+
+    ///
+    /// A datatype for pattern matching on a chain (traversing right-to-left).
+    ///
+    pub enum ViewRight[t] with Eq {
+        case NoneRight
+        case SomeRight(Chain[t], t)
+    }
+
+    ///
+    /// Return the empty chain.
+    ///
+    pub def empty(): Chain[a] = Empty
+
+    ///
+    /// Return the singleton chain with element `x`.
+    ///
+    pub def singleton(x: a): Chain[a] = One(x)
+
+    ///
+    /// Apply every function from `f` to every argument from `x` and return a chain with all results.
+    /// For `f = f1, f2, ...` and `x = x1, x2, ...` the results appear in the order
+    /// `f1(x1), f1(x2), ..., f2(x1), f2(x2), ...`.
+    ///
+    pub def ap(f: Chain[a -> b & ef], x: Chain[a]) : Chain[b] & ef =
+        apHelper(viewLeft(f), x, empty())
+
+    ///
+    /// Helper function for `ap`.
+    ///
+    def apHelper(f: ViewLeft[a -> b & ef], x: Chain[a], acc: Chain[b]): Chain[b] & ef = match f {
+        case NoneLeft         => acc
+        case SomeLeft(f1, rs) => apHelper(viewLeft(rs), x, acc `append` map(f1, x))
+    }
+
+    ///
+    /// Returns true if and only if `xs` is the empty chain, i.e. `Empty`.
+    ///
+    pub def isEmpty(xs: Chain[a]): Bool = match xs {
```

Comment:
I notice that the chains is often called `x` and `xs`. We used to use the same convention for list. 

The most reason idea is that `x` and `xs` inside a function (e.g. pattern match) is fine, but that we now prefer to name the arguments after the name of the "thing". E.g. `l` for `List`, `s` for `Set`, `m` for `Map`. I think here we should always use `c` for `Chain`.

Reply:
(I know that x and xs is standard in the FP community, but I am trying to be a bit more palatable to new programmers and to non-FP programmers, and so I think `c` is better than `xs`).

Reply:
All should now be changed to `c`.

Where there was an `xs` and `ys` like `eq` they should be `c1` and `c2`. Views should be `v`.